### ISSUE TITLE
Double wildcard in string query causes incorrect highlighting for plain and fast vectors highlighters

### ISSUE THREAD

**Issue by luckywin** [elastic/elasticsearch#17500]
**Elasticsearch version**: 1.5.2, 2.2.1

**JVM version**: java version "1.8.0_60" (Java HotSpot(TM) 64-Bit Server VM (build 25.60-b23, mixed mode))

**OS version**: Windows 7 Enterprize Service Pack 1

**Description of the problem including expected versus actual behavior**:
Expected same highlight results for query `*` and queries `**`, `*?` but in fact they are totally different.

**Steps to reproduce**:
Create following mapping:

```
PUT http://localhost:9200/index/_mapping/sometype 
{
    "properties" : {
        "sometext" : {
            "type" : "string",
            "term_vector" : "with_positions_offsets"
        }
    }
}
```

Post data

```
POST http://localhost:9200/index/sometype
{
    "sometext" : "A supervisor is responsible for the productivity and actions of a small group of     employees. The supervisor has several manager-like roles, responsibilities, and powers. Two of the key differences between a supervisor and a manager are (1) the supervisor does not typically have hire and fire authority, and (2) the supervisor does not have budget authority."
}
```

Query data with highlithing

```
POST http://localhost:9200/index/sometype/_search
{
    "query" : {
        "query_string" : {
            "query" : "**",
            "fields" : ["sometext"]
        }
    },
    "highlight" : {
        "pre_tags" : ["<em>"],
        "post_tags" : ["</em>"],
        "order" : "score",
        "require_field_match" : true,
        "fields" : {
            sometext : {
                "fragment_size" : 150,
                "number_of_fragments" : 1
            }
        }
    }
}
```

and you will got following (incorrect) highlighting:

```
"highlight" : {
    "sometext" : ["responsibilities, <em>and</em> <em>powers</em>. <em>Two</em> <em>of</em> <em>the</em> <em>key</em> <em>differences</em> <em>between</em> <em>a</em> <em>supervisor</em> <em>and</em> <em>a</em> <em>manager</em> <em>are</em> (<em>1</em>) <em>the</em> <em>supervisor</em> <em>does</em> <em>not</em> <em>typically</em> <em>have</em> <em>hire</em> <em>and</em> <em>fire</em> <em>authority</em>, and"]
}
```

The same highlighting results are produced by query `*?`.
On plain highlighter (I just added `"type" : "plain"` to highlight) result looks a bit different (but still incorrect):

```
"highlight" : {
    "sometext" : [", <em>responsibilities</em>, <em>and</em> <em>powers</em>. <em>Two</em> <em>of</em> <em>the</em> <em>key</em> <em>differences</em> <em>between</em> <em>a</em> <em>supervisor</em> <em>and</em> <em>a</em> <em>manager</em> <em>are</em> (<em>1</em>) <em>the</em> <em>supervisor</em> <em>does</em> <em>not</em> <em>typically</em> <em>have</em> <em>hire</em> <em>and</em> <em>fire</em> <em>authority</em>, <em>and</em> (<em>2</em>) <em>the</em> <em>supervisor</em> <em>does</em> <em>not</em> <em>have</em> <em>budget</em> <em>authority</em>."]
}
```

But when query is `*` - nothing returned by highlighter.

**Comment by clintongormley** [comment]
`*` is mapped to a match-all query, so nothing is highlighted.  `**` and `*?` (and `?*` and `***` and `*?*?`) end up doing a wildcard query on all terms, so all terms are highlighted.  Not sure what we can really do about this.

**Comment by nik9000** [comment]
It might be worth looking at * just doing a wildcard match all for text and making sure that is rewritten sensibly. You can certainly use `highlight_query` in the mean time if you find searching for `**` to be slower than `*`. At a minimum we need to update the docs page for highlighting.

The "incorrect" highlighting you describe is actually the expected behavior of the fast vector highlighter. It identifies hits, picks the right snippet based on the hits, and then expands the snippet forwards and backwards looking for word breaks. It doesn't go back to the list of hits when it does the seeking. IIRC the experimental-highlighter did do that, but no one has updated it to work with 2.0 so, yeah.

**Comment by clintongormley** [comment]
Closed in favour of #21621
